synapse_user_restrictions/config.py

```diff
@@ -0,0 +1,127 @@
+import abc
+import enum
+import re
+from enum import Enum
+from typing import Any, List, Pattern, Set, Type
+
+import attr
+import cattr
+from attr import Attribute
+from cattr import Converter
+
+
+def make_cattr_converter() -> Converter:
+    """
+    Creates a cattr `Converter` instance, with custom registered hooks for
+    regular expressions.
+    """
+    converter = cattr.Converter()
+
+    def regex_structure_hook(value: Any, _kind: Type[Pattern[str]]) -> Pattern[str]:
+        if not isinstance(value, str):
+            raise ValueError("Regex patterns should be specified as strings.")
+
+        try:
+            return re.compile(value)
+        except re.error as e:
+            raise ValueError(f"Invalid regex '{value}': {e.msg}")
+
+    converter.register_structure_hook(Pattern[str], regex_structure_hook)
+
+    return converter
+
+
+def check_all_permissions_understood(
+    _instance: Any, _attribute: "Attribute[Set[str]]", permissions: Set[str]
```

Comment:
Isn't the `instance` of type `UserRestrictionsModuleConfig`?

I'm surprised the `Attribute[Set[str]]` needs quotes, but not a big deal!

Isn't `permissions` actually of type `Any`?